### PR TITLE
Add shape and volume builder

### PR DESCRIPTION
**Pull request by sethrj**
Follow-on to #1125

This defines the beginning of the `Object` hierarchy, the `VolumeBuilder` helper class for constructing volumes from a shape, and the `Shape` class. There will also be `Transformed`, `Solid`, and other implementations of the `ObjectInterface`.

```mermaid
classDiagram
  Object <|-- Shape
  ShapeBase <|-- Shape
  class Object {
    +string_view label()*
    +NodeId build(VolumeBuilder&)*
  }
  <<Interface>> Object
  class ShapeBase {
    #void build_interior(ConvexSurfaceBuilder&)*
  }
  <<Abstract>> ShapeBase
  class Shape {
    -string label;
    -ConvexRegion region;
  }
  Shape *-- ConvexRegion
  class ConvexRegion {
    +void build(ConvexSurfaceBuilder&)*
  }
  <<Interface>> ConvexRegion
  ConvexRegion <|-- Box
  ConvexRegion <|-- Sphere
  class Box {
    -Real3 halfwidths
  }
  class Sphere {
    -real_type radius
  }

  Shape <|.. BoxShape
  Shape <|.. SphereShape

  BoxShape *-- Box
  SphereShape *-- Sphere
```

### CODE REVIEW

**Comment by sethrj** [comment]
Thanks for the reviews @elliottbiondo ! This one is next.

File: src/orange/orangeinp/Shape.hh [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+/*!
+ * A simple, convex region of space.
+ *
+ * This is an abstract class that implements \c build for constructing a volume
+ * by dispatching to a method \c build_interior that the daughters must
+ * override using a convex region.
+ *
+ * Use the implementation classes \c XShape where \c X is one of the convex
+ * region types in ConvexRegion.hh :
+ * - \c BoxShape
```
Comment (elliottbiondo) [inline src/orange/orangeinp/Shape.hh:29]: Since it's XShape you can just say Box, Cone, Cylinder, etc

**Review by elliottbiondo — APPROVED** [review]
There are quite a few levels of abstraction here, which I understand is necessary to build geometries from different sources. At some point it would be nice to have a brief, high-level design diagram/readme/document referenced within the orangeinp dir concisely summarizing what goes on in each step from convex surfaces -> region -> shape -> object -> (csg unit) -> volume. This would also be helpful because I think there are some minor naming clashes, such as "unit" within both CSGUnits, and universe units. Obviously these are very different things, but they could be easily mixed up if one does not know better.

File: src/orange/orangeinp/Shape.hh (hunk 1 above)
Comment (sethrj) [reply src/orange/orangeinp/Shape.hh:29]: True, but I wanted to use the exact type alias name to elevate the type aliases which have to be at the bottom of the file. I'd hoped also that the doxygen would generate links from those exact names, but it seems not to ...

**Comment by sethrj** [comment]
Thanks! I will definitely add some diagrams once this gets to the "build a unit/scene" phase.

**Pull request merged**